### ISSUE TITLE
CrossEntropy yields negative or NaN loss without a probability output layer

### ISSUE THREAD

**Issue by lminervino18** [lminervino18/oxidized-neural-orchestra#171]
## Resumen
`CrossEntropy` no aplica softmax internamente ni valida que sus entradas sean probabilidades. Cuando la última capa no es una distribución de probabilidad, el loss se vuelve **negativo** o **diverge a NaN**.

## Evidencia
| activación final | resultado |
|---|---|
| `softmax` | OK |
| `sigmoid` | OK (acotado 0–1) |
| `relu` | **loss negativo** (ej. `-0.198`) |
| `tanh` | basura / no converge (~`5.3`) |
| lineal (ninguna) | **NaN** -> `worker error: loss diverged: NaN or Inf` |

El NaN se detecta y la corrida aborta limpiamente, pero el **loss negativo es silencioso y matemáticamente inválido**.

## Configuración que lo reproduce (FFI)
```python
# la capa final no tiene softmax/sigmoid
model = Sequential([
    Dense(8, Kaiming(), ReLU()),
    Dense(3, Kaiming(), ReLU()),
])
training = orchestra.parameter_server(
    addrs=[...], nservers=1, dataset=dataset,
    optimizer=Adam(0.005), loss_fn=CrossEntropy(),   # CE sobre activaciones crudas
    sync=BarrierSync(), store=BlockingStore(),
    max_epochs=60, batch_size=32,
)
# -> el avg_loss se vuelve negativo; con salida lineal diverge a NaN
```

## Fix sugerido
Integrar el softmax dentro de `CrossEntropy` (softmax-cross-entropy), o validar/acotar las entradas a un rango de probabilidad válido.

**Comment by alejoordonez02** [comment]
esto es una feature no un bug! (dea) 
lo que pasa es que hay un quilombo me parece a mí con los nombres. CrossEntropy no debería recibir input logits, los logits son los elementos del vector y_pred que no están normalizados en una distribución de probabilidad; es una loss para comparar justamente distribuciones de probabilidad, entonces hay que usarlo con algo que ya te haya aplanado los y_preds.
Lo que hace PyTorch es: se olvida de esta cuestión definitoria y directamente hace softmax adentro de la loss. Por qué? porque la cuenta compuesta (o sea cross_entropy(softmax(x))) es una resta. Pero técnicamente no está haciendo cross entropy, es una SoftmaxCrossEntropy ponele.
Voy a ver si conviene volar esta formalidad o si implementar la SoftmaxCrossEntropy loss, que espera logits y no distribuciones de probabilidad.

**Comment by alejoordonez02** [comment]
Lo estuve mirando y hasta ahora no se me ocurrió ni vi una forma de hacerlo que no implique básicamente mover softmax adentro de crossentropy para después hacer la resta que digo. No sé si cambia demasiado...
